python/ql/src/experimental/semmle/python/security/injection/XSLT.qll

```diff
@@ -0,0 +1,114 @@
+/**
+ * Provides class and predicates to track external data that
+ * may represent malicious XSLT query objects.
+ *
+ * This module is intended to be imported into a taint-tracking query
+ * to extend `TaintKind` and `TaintSink`.
+ */
+
+import python
+import semmle.python.dataflow.TaintTracking
+import semmle.python.web.HttpRequest
+
+/** Models XSLT Injection related classes and functions */
+module XSLTInjection {
+  /** Returns a class value which refers to `lxml.etree` */
+  Value etree() { result = Value::named("lxml.etree") }
+
+  /** A generic taint sink that is vulnerable to XSLT injection. */
+  abstract class XSLTInjectionSink extends TaintSink { }
+
+  /**
+   * A kind of "taint", representing an untrusted XML string
+   */
+  private class ExternalXmlStringKind extends ExternalStringKind {
+    ExternalXmlStringKind() { this = "etree.XML string" }
+
+    override TaintKind getTaintForFlowStep(ControlFlowNode fromnode, ControlFlowNode tonode) {
+      etreeXML(fromnode, tonode) and result instanceof ExternalXmlKind
+      or
+      etreeFromStringList(fromnode, tonode) and result instanceof ExternalXmlKind
+      or
+      etreeFromString(fromnode, tonode) and result instanceof ExternalXmlKind
+    }
+  }
+
+  /**
+   * A kind of "taint", representing a XML encoded  string
+   */
+  class ExternalXmlKind extends TaintKind {
+    ExternalXmlKind() { this = "lxml etree xml" }
+  }
+
+  private predicate etreeXML(ControlFlowNode fromnode, CallNode tonode) {
+    exists(CallNode call | call.getFunction().(AttrNode).getObject("XML").pointsTo(etree()) |
+      call.getArg(0) = fromnode and
+      call = tonode
+    )
+  }
+
+  private predicate etreeFromString(ControlFlowNode fromnode, CallNode tonode) {
+    // fromstring(text, parser=None)
```

Comment:
I think it would be slightly clearer to qualify the call in the comment:
`// etree.fromstring(test, parser=None)`